torchmdnet/models/utils.py

```diff
@@ -76,6 +77,143 @@ def forward(self, z, x, edge_index, edge_weight, edge_attr):
     def message(self, x_j, W):
         return x_j * W
 
+import torchmdnet.neighbors as neighbors
+class DistanceCellList(torch.nn.Module):
+
+    def __init__(
+        self,
+        cutoff_lower=0.0,
+        cutoff_upper=5.0,
+        max_num_pairs=-32,
+        return_vecs=False,
+        loop=False,
+        strategy="brute",
+        include_transpose=True,
+        resize_to_fit=True,
+        check_errors=True,
+        box=None,
+    ):
+        super(DistanceCellList, self).__init__()
+        """ Compute the neighbor list for a given cutoff.
+        This operation can be placed inside a CUDA graph in some cases.
+        In particular, resize_to_fit and check_errors must be False.
+        Note that this module returns neighbors such that distance(i,j) >= cutoff_lower and distance(i,j) < cutoff_upper.
+        Parameters
+        ----------
+        cutoff_lower : float
+            Lower cutoff for the neighbor list.
+        cutoff_upper : float
+            Upper cutoff for the neighbor list.
+        max_num_pairs : int
+            Maximum number of pairs to store.
+            If negative, it is interpreted as (minus) the maximum number of neighbors per atom.
+        strategy : str
+            Strategy to use for computing the neighbor list. Can be one of
+            ["shared", "brute", "cell"].
+            Shared: An O(N^2) algorithm that leverages CUDA shared memory, best for large number of particles.
+            Brute: A brute force O(N^2) algorithm, best for small number of particles.
+            Cell:  A cell list algorithm, best for large number of particles, low cutoffs and low batch size.
+        box : Optional[torch.Tensor]
+            Size of the box, shape (3,3) or None.
+            If strategy is "cell", the box must be diagonal.
+        loop : bool
+            Whether to include self-interactions.
+        include_transpose : bool
+            Whether to include the transpose of the neighbor list.
```

Comment:
What is this?

Reply:
The current Distance module includes both the (i,j) and (j,i) pairs. This module allows to omit the redundant pairs. I understood this as "including the transpose".

Reply:
And loop is how Distance (and pytorch_cluster) calls "add particles as neighbors of themselves". I am not sure where this name comes from, but to keep it consistent...